### PR TITLE
[GEOS-11207] Refactor MapML MVC controller as GetMap-based operation with standard parameter format

### PR DESCRIPTION
**Pull request by turingtestfail**
MapML controller refactored as a WMS Output format.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by prushforth** [comment]
A few comments, I will continue to evaluate and add to them I can.  Might take a few days. Also, the mapml-viewer doesn't seem to work to query multiple \<map-extent> elements, but a more recent build of our current work does appear to work better for that use case.  I will switch back and forth, but I am only worrying about the server functionality for your work.  We'll update the viewer when there's an opportunity.  Also, I would be willing to try to offer some edits of your PR if necessary.  Just let me know if you feel that there's something I mention that I should tackle.

- [ ]  multi-layer, single \<map-extent> (set via default WMS global MapML rendering setting) preview request does not appear to include a \<map-link rel=query> [example](http://localhost:8080/geoserver/ows/wms?service=WMS&version=1.1.0&request=GetMap&layers=topp%3Astates,spearfish&bbox=-124.73142200000001%2C24.955967%2C-66.969849%2C49.371735&width=768&height=330&srs=EPSG%3A4326&styles=&format=text/html;%20subtype=mapml), but a single-layer \<map-extent> preview request does include the query link - [example](http://localhost:8080/geoserver/ows/wms?service=WMS&version=1.3.0&request=GetMap&layers=topp%3Astates,spearfish&bbox=-124.73142200000001%2C24.955967%2C-66.969849%2C49.371735&width=768&height=330&srs=EPSG%3A4326&styles=&format=text/html;%20subtype=mapml)
- [ ] multi-\<map-extent> GetMap, query working well (with newer version of the mapml-viewer) [example (same as previous link with WMS global MapML rendering setting changed to emit multiple \<map-extent>s)](http://localhost:8080/geoserver/ows/wms?service=WMS&version=1.1.0&request=GetMap&layers=topp%3Astates,spearfish&bbox=-124.73142200000001%2C24.955967%2C-66.969849%2C49.371735&width=768&height=330&srs=EPSG%3A4326&styles=&format=text/html;%20subtype=mapml)
- [ ] for multi-\<map-extent> GetMap requests, the hidden="hidden" attribute could be omitted from the \<map-extent> elements.  The purpose of multi-\<map-extent> layers is to enable the user to turn them on / off via the layer control.  `hidden="hidden"` \<map-extent> and \<layer-> elements are not presented in the layer control, by design.
- [ ] the default MapML preview link does render the layer at the center of the map, it seems only for EPSG:4326 URLs?  I imagine that the bbox that is calculated for the preview link URL is the bounds of the layer in the coordinate system of the preview, and I also imagine that the lat/lon/zoom values of the \<mapml-viewer> in the preview is set based on that bbox.  So I guess it's pretty difficult to 'hack' the URL to get a good preview.  If you right-click the layer in the layer control and select 'Zoom to layer', you get a good preview, it seems.
- [ ] Some preview links don't work because the srs referenced doesn't have a MapML TCRS mapping.  Could we change the preview URL to a standard TCRS-mappable EPSG code? [example](http://localhost:8080/geoserver/sf/wms?service=WMS&version=1.1.0&request=GetMap&layers=sf%3Aroads&bbox=589434.8564686741%2C4914006.337837095%2C609527.2102150217%2C4928063.398014731&width=768&height=537&srs=EPSG%3A26713&styles=&format=text/html;%20subtype=mapml)
- [ ] alternate projection links are invalid geoserver WMS links   - when dereferenced we get:

> Missing or invalid requested map size. Parameters WIDTH and HEIGHT shall be present and be integers > 0. Got WIDTH=0, HEIGHT=0

**Comment by turingtestfail** [comment]
Peter,
Changes just pushed.  Some notes on your notes:
1) Fixed to include query links for a multi-layer request.
2) Glad to hear it.
3) Hidden removed from mult- map-extent.
4) Still looking into this.
5) I believe we will be discussing the way forward for TCRS at the next
meeting?
6) Alternate project links fixed to include WIDTH and HEIGHT.

Thanks,
Joe Miller

**Comment by prushforth** [comment]
- Only EPSG:4326 previews work; you can't get a correct preview for a layer set up with EPSG:3857, EPSG:3978, EPSG:5936 as the  declared EPSG.  To try to get the USA Population preview to display in EPSG:3857, for example, I used the following settings:
image


The when the MapML preview link is generated, the map loads but is blank / empty.  If I right-click on the layer in the layer control and select "Zoom to layer", the preview is correctly displayed.  This suggests that the starting location of the \<mapml-viewer lat=... lon=... zoom=...> element in the HTML page is not being correctly set to the de-projected center of the layer bounds.  Upon examination, the lat and lon appear to contain non-geographic coordinates initially.

**Comment by prushforth** [comment]
Hi Joe,

> 6) Alternate project links fixed to include WIDTH and HEIGHT.

The bbox parameter is missing from the alternate projection links.  Sorry I should have picked up on this before, but the error message only mentioned width and height at first, but now it's mentioning bbox.

**Comment by prushforth** [comment]
Joe, 

Another thing: 

In the MapML viewer, you can select an alternate style from the Layer settings menu, under "Style":

image

Note that the label on the default style above is empty because the \<map-link> served is missing the `title` attribute: `<map-link href="http://.../geoserver/wms?format_options=mapml-wms-format%3Aimage%2Fpng&amp;request=GetMap&amp;crs=EPSG%3A4326&amp;service=WMS&amp;format=text%2Fmapml&amp;layers=states&amp;styles=&amp;version=1.3.0"
            rel="self style" title="" />`

The \<map-link rel="style" ...> and \<map-link rel="self style"> URLs also don't have the BBOX nor WIDTH/HEIGHT parameters.  Note that in the old MapML extension, I found that I had to ensure that the default WMS style was also added to the "Selected Styles" list box in order to ensure that each of the links to an alternate style "knew about" (linked to) all available styles, including the default style... I forget why that was the case.

image

Defaults style added to "Selected Styles", required by MapML extension to have complete set of links to all available styles:

image

Even when you add the population style to the "Available Styles", it's still blank / empty string in the document:

image

**Comment by prushforth** [comment]
Hi Joe and Andrea,

I noticed that the MapML \<map-extent> content is always configured to do GetMap requests, regardless of whether there's a GeoWebCache GridSet for it or not.  On main, the MapML extension checks the "Use Tiles" value of the Layer's/Layer Group's MapML Settings panel under "Publishing", and if it's checked, it will check to see if a GridSet is configured for the Layer under that projection. If so, it serializes the \<map-extent> content to use WMTS GetTile requests instead of tile-shaped WMS GetMap requests.  I realize that it's not "refactoring" to build that functionality for multi-layer GetMap requests, but maybe it should continue to work for single-layer requests and potentially multi-layer/multi-layergroup requests for which the WMS MapML Settings value is "Represent multi-layer requests as multiple elements" == true.

Do you think it will be possible to carry forward this functionality within the scope of our current work?  Should it wait until we have dealt with TCRS implementation?

Thank you!

**Comment by turingtestfail** [comment]
Peter,
I just pushed the following:
1)Added BBOX, HEIGHT, WIDTH to Alt, Style, and Self-Style links.
2)Changed the HTML/Preview capability so that the center lat and long are
always WGS84 (this should fix the issue you saw with non-WGS84 previews)
3)For Self-Style, in the event that no style is specified in the request it
tries to build a comma delimited list of default styles.  Note that in the
case of Layer Groups it checks for the first "Layer Group Style".

Joe

**Comment by prushforth** [comment]
Hi Joe,

> 1)Added BBOX, HEIGHT, WIDTH to Alt, Style, and Self-Style links.

It seems that the `styles=` WMS parameter is empty on all the \<map-link rel="style" or \<map-link rel="self style" links, though the `title` attribute is now set in all cases.  I see that you just pushed something will check it out.  The link should lead to the mapml document that yields the named style fwiw.

**Comment by turingtestfail** [comment]
Yeah, I noticed that I was setting the title for self-style but not the
styles parameter.  I also just made another push that addresses the same
issue for the other style links.

Joe

**Comment by prushforth** [comment]
By the way, nice job getting rid of all that namespace prefix cruft.  Amazing work!

**Comment by prushforth** [comment]
Sorry for the trouble, but getting an NPE on mult-layer requests like [this one](http://localhost:8080/geoserver/ows/wms?service=WMS&version=1.1.0&request=GetMap&layers=topp%3Astates,spearfish&bbox=-1.3885038382960921E7%2C2870337.130793682%2C-7455049.489182421%2C6338174.0557576185&width=768&height=414&srs=EPSG%3A3857&styles=&format=text/html;%20subtype=mapml)

**Comment by prushforth** [comment]
> Sorry for the trouble, but getting an NPE on mult-layer requests like [this one](http://localhost:8080/geoserver/ows/wms?service=WMS&version=1.1.0&request=GetMap&layers=topp%3Astates,spearfish&bbox=-1.3885038382960921E7%2C2870337.130793682%2C-7455049.489182421%2C6338174.0557576185&width=768&height=414&srs=EPSG%3A3857&styles=&format=text/html;%20subtype=mapml)

Note that I can't reproduce the NPE so hiding the comment as resolved...

**Comment by prushforth** [comment]
Joe,

If acceptable, I could push a small update and associated assertion in MapMLWMSTest.java which sets the `label` attribute on \<map-extent> when multi-extent WMS setting is enabled?  If you don't like it you can remove the commit?

Thanks

**Comment by turingtestfail** [comment]
Apologies for the late response.  All updates are welcome!  I'll take a
look and merge tomorrow.

Thanks,
Joe

**Comment by prushforth** [comment]
The `styles` parameter isn't quite working for multi-layer requests.  For example, if I provide a `layers=Img_Sample,states,spearfish&styles=,pophatch,`, I get a java.lang.ArrayIndexOutOfBoundsException at org.geoserver.mapml.MapMLDocumentBuilder.initialize(MapMLDocumentBuilder.java:266)

**Comment by turingtestfail** [comment]
Peter,
I just pushed a fix for the issue you found.  The array split I was using
was ignoring the trailing empty style.  It should now treat that as an
empty string.  If there are fewer styles than layers it will also now
return a more informative error message.

 Apologies for the miss,
Joe

**Comment by prushforth** [comment]
>  Apologies for the miss,

No worries, everything is up for evaluation. But we should put a test on it.

**Comment by prushforth** [comment]
It's interesting that in the previous version, all resources that matched the pattern "text/.*" would be passed through the gzip filter and they would take the full `Keep-Alive: timeout=20` header value of 20 seconds to complete the request in the browser.  So that would include the `text/html` and `text/mapml` responses from the original MapMLController.  

In the current version, there is no MapMLController, and the `text/html; subtype=mapml` and `text/mapml` responses are gzipped and only take a few milliseconds to complete.  However, the resource file "mapml.css" (served as `text/css`) still takes the full 20 seconds.

(The reason these are taking 20 seconds is because the Content-Length: 28502 value is set to the unzipped file size, not the zipped content size, so the client leaves the connection open for the full 20 seconds of the Keep-Alive value)

So whatever process is passing the resource file to the gzip filter is not working well with the gizp filter, and I guess that process is not the same process as the WMS request processor.

Long story short, the user still has to remove the `text/.*` from the web.xml gzip filter parameter.

**Comment by prushforth** [comment]
Caveat for [this commit](https://github.com/geosolutions-it/geoserver/pull/354/commits/4379932534e1076e17eb8eea6bc3098888238fe2) is that if you want the default style correctly listed, you need to explicitly add a copy of the base layer group configuration as a layer group style named "default-style-"+layerGroupName (which is the name assigned to the default in the capabilities anyway). Trouble is that it ends up as a duplicate in the capabilities, but it mirrors the same difficulty with default styles for layers, which is that you must add the default style as an explicitly added style (in the case of layer groups, you must name it "default-style-"+layerGroupName, resulting in the duplication in the capabilities document).

Long story short, I need to add a test or two for this, and I should add documentation for how to get access to named styles for both layers and layer groups.
